illumos-utils/src/dladm.rs

```diff
@@ -174,10 +171,121 @@ impl VnicSource for PhysicalLink {
+        let mut command = std::process::Command::new(PFEXEC);
+        let mut args = vec![
+            DLADM.to_string(),
+            "create-vnic".to_string(),
+            "-t".to_string(),
+            "-l".to_string(),
+            source.name().to_string(),
+        ];
+
+        if let Some(mac) = mac {
+            args.push("-m".to_string());
+            args.push(mac.0.to_string());
+        }
+
+        if let Some(vlan) = vlan {
+            args.push("-v".to_string());
+            args.push(vlan.to_string());
+        }
+
+        args.push("-p".to_string());
+        args.push(format!("mtu={mtu}"));
+
+        args.push(vnic_name.to_string());
+
+        let cmd = command.args(&args);
+        execute(cmd).map_err(|err| CreateVnicError {
+            name: vnic_name.to_string(),
+            link: source.name().to_string(),
+            err,
+        })?;
+
+        // In certain situations, `create-vnic -p mtu=N` does not actually set
+        // the mtu to N. Set it here using `set-linkprop`.
+        //
+        // See https://www.illumos.org/issues/15695 for the illumos bug.
+        let mut command = std::process::Command::new(PFEXEC);
+        let prop = format!("mtu={}", mtu);
+        let cmd = command.args(&[
+            DLADM,
+            "set-linkprop",
+            "-t",
+            "-p",
+            &prop,
+            vnic_name,
+        ]);
+        execute(cmd).map_err(|err| CreateVnicError {
+            name: vnic_name.to_string(),
+            link: source.name().to_string(),
+            err,
+        })?;
+
+        Ok(())
+    }
+
+    /// Verify that the given link exists
+    fn verify_link(&self, link: &str) -> Result<Link, FindPhysicalLinkError> {
+        let mut command = std::process::Command::new(PFEXEC);
+        let cmd = command.args(&[DLADM, "show-link", "-p", "-o", "LINK", link]);
+        let output = execute(cmd)?;
+        match String::from_utf8_lossy(&output.stdout)
+            .lines()
+            .next()
+            .map(|s| s.trim())
+        {
+            Some(x) if x == link => Ok(Link::wrap_physical(link)),
+            _ => Err(FindPhysicalLinkError::NoPhysicalLinkFound),
+        }
+    }
+
+    /// Remove a vnic from the sled.
+    fn delete_vnic(&self, name: &str) -> Result<(), DeleteVnicError> {
+        let mut command = std::process::Command::new(PFEXEC);
+        let cmd = command.args(&[DLADM, "delete-vnic", name]);
+        execute(cmd)
+            .map_err(|err| DeleteVnicError { name: name.to_string(), err })?;
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl Api for Dladm {}
 
-#[cfg_attr(any(test, feature = "testing"), mockall::automock, allow(dead_code))]
 impl Dladm {
+    /// Access the real dladm API, which will invoke commands on the host OS.
+    ///
+    /// If you're interested in testing this interface, consider using
+    /// [crate::fakes::dladm::Dladm] instead.
+    pub fn real_api() -> Self {
+        Self { phantom: PhantomData }
```

Comment:
This comment applies here, and to the same function in `zone.rs`:

I'm using the `real_api()` as a name, and `PhantomData`, to make it impossible for callers to create a `Dladm` object via `Dladm {}`. 

This is for readability: I want it to be extremely clear when a "real_api" vs "fake_api" is being used.

Reply:
Useless code golf: if you want to avoid pulling in `PhantomData` I think you can also define this as `struct Dladm(())` and get the same effect, but this works fine too.

Reply:
Good call, done in 1add1894d6c1b1437102e53e78b94b07b26df675